### PR TITLE
Add segment module

### PR DESCRIPTION
**Pull request by roshankern**
@gwaybio This is ready for review.

What relevant CPU/GPU specs should we include in the README? Right now we are just copying the output of `nvidia-smi` for GPU specs and not including any CPU specs (as the code runtime is significantly GPU-dependent).

### CODE REVIEW

File: 1.segment-data/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+We use the CellPose segmentation algorithm to segment the nuclei from the Hoechst channel of each Cell Painting image.
+CellPose was first introduced in [Stringer, C., Wang, T., Michaelos, M. et al., 2020](https://doi.org/10.1038/s41592-020-01018-x) and we use the [python implementation](https://github.com/mouseland/cellpose).
+
+Stringer et al. trained the CellPose segmentation models on a diverse set of cell images and is therefore a good selection for our use case.
+The CellPose python implementation was particularly useful for building reproducible pipelines.
+
+### Nuclei segmentation
+
+After manually experimenting with CellPose on about 10 Cell Health nuclei images (various cell lines), we settled on the following parameters for CellPose nuclei segmentation:
```
Comment (gwaybio) [inline 1.segment-data/README.md:15]: optional, but it might be nice to elaborate a bit more on what you mean by "manually experimenting" - what were your eyes drawn to in determining how effective different configurations were?

File: 1.segment-data/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+- `model_type = "cyto"` This parameter forces CellPose to use the cytoplasm model, which we found segments nuclei in the Cell Health data significantly better than the nucleus model.
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
- `model_type = "cyto"` This parameter forces CellPose to use the cytoplasm model, which we found segments nuclei in the Cell Health data substantially better than the nucleus model. 
```

Only use "significant" when performing a statistical test :)

File: 1.segment-data/README.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+### Cytoplasm segmentation
+
+After manually experimenting with CellPose on about 10 Cell Health cytoplasm images (various cell lines), we settled on the following parameters for CellPose cytoplasm segmentation:
+- `model_type = "cyto"` This parameter forces CellPose to use the cytoplasm model for segmentation.
+More information about CellPose models can be found at https://cellpose.readthedocs.io/en/latest/models.html.
+- `channels = [1,3]` This parameter forces the model to segment cells using the nuclei and RNA channels of cell painting images. 
+We overlay nuclei, ER, and RNA channels for CellPose with our `overlay_images()` function in [segment-cell-health-data.ipynb](segment-cell-health-data.ipynb).
+
+**Note:** The channel numbers 1,3 correspond to the CellPose colors red and blue respectively.
```
Comment (gwaybio) [inline 1.segment-data/README.md:42]: Great to see this elaboration - can you add a bit more about ER? This is a non-standard approach (innovative!) and I think it's worth mentioning a bit more detail (I don't see ER mentioned in the note)

File: 1.segment-data/README.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+The CellPose python implementation was particularly useful for building reproducible pipelines.
+
+### Nuclei segmentation
+
+After manually experimenting with CellPose on about 10 Cell Health nuclei images (various cell lines), we settled on the following parameters for CellPose nuclei segmentation:
+- `model_type = "cyto"` This parameter forces CellPose to use the cytoplasm model, which we found segments nuclei in the Cell Health data significantly better than the nucleus model. 
+More information about CellPose models can be found at https://cellpose.readthedocs.io/en/latest/models.html.
+- `channels = [0,0]` This parameter forces the model to segment cells in grayscale (in the case of Cell Health data single channel images).
+- `diameter = 80` This parameter indicates to the model that the average cell diameter is 80 pixels.
+- `flow_threshold=0` This paramenter decreases the maximum allowed error of the flows for each mask (default is `flow_threshold=0.4`).
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
- `flow_threshold = 0` This paramenter decreases the maximum allowed error of the flows for each mask (default is `flow_threshold = 0.4`).
```

for consistency

File: 1.segment-data/README.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+- `channels = [1,3]` This parameter forces the model to segment cells using the nuclei and RNA channels of cell painting images. 
+We overlay nuclei, ER, and RNA channels for CellPose with our `overlay_images()` function in [segment-cell-health-data.ipynb](segment-cell-health-data.ipynb).
+
+**Note:** The channel numbers 1,3 correspond to the CellPose colors red and blue respectively.
+The `overlay()` function makes RNA the red channel of the image and DNA the blue channel of the image.
+Thus, `channels = [1,3]` forces CellPose to segment the RNA channel using the DNA channel as its base.
+
+- `diameter = 0` This parameter forces the CellPose model to estimate the diameter of cells being segmented.
+- `flow_threshold=0` This paramenter decreases the maximum allowed error of the flows for each mask (default is `flow_threshold=0.4`).
+- `cellprob_threshold=0.4` This parameter is used to determine ROIs (default is `cellprob_threshold=0`).
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
- `flow_threshold = 0` This paramenter decreases the maximum allowed error of the flows for each mask (default is `flow_threshold = 0.4`).
- `cellprob_threshold = 0.4` This parameter is used to determine ROIs (default is `cellprob_threshold = 0`).
```

For consistency

File: 1.segment-data/README.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+```
+
+### Step 1c (Optional): Complete PyTorch GPU Setup
+
+If you would like use PyTorch GPU when using CellPose, follow [these instructions](https://github.com/MouseLand/cellpose#gpu-version-cuda-on-windows-or-linux) to complete the PyTorch GPU setup.
+We use PyTorch GPU while segmenting Cell Health data.
+
+## Step 2: Define Data Paths
+
+`load_path` and `save_path` need to be changed to reflect the desired load/save locations.
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
Inside the notebook <INSERT_NAME_HERE> the variables`load_path` and `save_path` need to be changed to reflect the desired load/save locations.
We used an external harddrive and therefore needed to use specific paths.
```

File: 1.segment-data/README.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+## Step 3: Execute Training Data Segmentation
+
+```bash
+# Run this script to segment Cell Health data
+bash 1.segment-data.sh
+```
+**Note:** With GPU enabled, our estimated run times were as follows:
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion

**Note:** With GPU enabled, our estimated run times were as follows:
```

Just adding an extra space after the bash code block

File: 1.segment-data/cell-health-segmentation.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        if current_image in image_file.name:
+            channel_paths.append(image_file)
+    channel_paths = sorted(channel_paths)
+    for channel_path in channel_paths:
+        channel_image = np.array(cv2.imread(str(channel_path), 0))
+        channel_images.append(channel_image)
+
+    # channel multipliers are only useful for human interpretation of the overlayed images
+    # channel brightness is irrelevant to CellPose segmentation
+    # channel images 1,2,3 are DNA, ER, RNA respectively
```
Comment (gwaybio) [inline 1.segment-data/cell-health-segmentation.py:109]: 🎉

File: 1.segment-data/segment-cell-health-data.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+from cellpose import models, core, io, utils
+
+import pathlib
+import pandas as pd
+
+import cv2
+import numpy as np
+
+import importlib
+chs = importlib.import_module("cell-health-segmentation")
```
Comment (gwaybio) [inline 1.segment-data/segment-cell-health-data.py:22]: neat! I've never seen this before

File: 1.segment-data/segment-cell-health-data.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+cytoplasm_model_specs = {
+    "model_type": "cyto",
+    "channels": [1,3],
+    "diameter": 0,
+    "flow_threshold": 0,
+    "cellprob_threshold": 0.4
+}
+
+chs.segment_cell_health(load_path, save_path, nuclei_model_specs, cytoplasm_model_specs)
```
Comment (gwaybio) [inline 1.segment-data/segment-cell-health-data.py:65]: beautiful 😍

**Review by gwaybio — APPROVED** [review]
I made mostly minor suggestions to the documentation. Everything else looks wonderful. Please merge once you address my comments (no need for me to look again)

**Comment by roshankern** [comment]
@gwaybio Made lots, lots of changes. Might be worth reviewing again. If not, let us know and we will merge.

File: 1.segment-data/README.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+More information about CellPose models can be found at https://cellpose.readthedocs.io/en/latest/models.html.
+- `channels : [0,0]` This parameter forces the model to segment cells in grayscale (in the case of Cell Health data single channel images).
+- `diameter : 80` This parameter indicates to the model that the average cell diameter is 80 pixels.
+- `flow_threshold : 0` This paramenter decreases the maximum allowed error of the flows for each mask (default is `flow_threshold : 0.4`).
+- `cellprob_threshold : 0` This parameter is used to determine ROIs (default is `cellprob_threshold : 0`).
+- `remove_edge_masks : True` This parameter removes any masks from CellPose that are touching an edge of the image.
+
+When manually experimenting with CellPose parameters to segment nuclei, we looked at nuclei shape/texture to determine how we believe it should be segmented.
+
+For example, in the image below, the nuclei is a continous shape with a continuous, flat texture (left).
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
For example, in the image below, the nucleus is a continuous shape with a continuous, flat texture (left).
```

File: 1.segment-data/README.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+Thus, this nuclei should be segmented as one nuclei.
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
Thus, these segmented nuclei should be segmented as one nucleus.
```

File: 1.segment-data/README.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+CellPose thinks the nuclei below is two nuclei when `diameter` is set to 0 (middle).
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
CellPose thinks the nucleus below is two nuclei when `diameter` is set to 0 (middle).
```

File: 1.segment-data/README.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+With our parameters, Cellpose correctly segments the nuclei (right).
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
With our parameters, Cellpose correctly segments the nucleus (right).
```

File: 1.segment-data/README.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+![Cytoplasm Compiled](imgs/cyto_compiled.png "Cytoplasm Compiled")
+
+### Cell Identification
+
+After finding the nuclei center coordinates and the cytoplasm outlines, we determine which nuclei and cytoplasms are related.
+First, we assign a `Cell_ID` to every cytoplasm.
+Then, we create a [matplotlib.path](https://matplotlib.org/stable/api/path_api.html) (polygon) from cytoplasm outlines.
+If any nuclei center coordinates are within the cytoplasm polygon, they are assigned the same `Cell_ID` as the cytoplasm.
+Any cytoplasm that does not have a nucleus associated with it is discarded.
+This assures that segmentation errors are controlled (every cytoplasm must have at least one nucleus).
```
Comment (gwaybio) [inline 1.segment-data/README.md]:
```suggestion
After finding the nuclei center coordinates and the cytoplasm outlines, we determine which nuclei and cytoplasms are related.
First, we assign a `Cell_ID` to every cytoplasm.
Then, we create a [matplotlib.path](https://matplotlib.org/stable/api/path_api.html) (polygon) from cytoplasm outlines.
If any nuclei center coordinates are within the cytoplasm polygon, they are assigned the same `Cell_ID` as the cytoplasm.
We discard any cytoplasm that does not have an associated nucleus.
This controls for segmentation errors (every cytoplasm must have at least one nucleus).
```

File: 1.segment-data/cell-health-segmentation.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+    
+    # iterate over all cytoplasm outlines
+    for cyto_index, cyto_row in cyto_outlines.iterrows():
+        cell_id = cyto_row["Cell_ID"]
+        
+        # create path (polygon) from outline to check if the nuclei are in this path
+        outline = cyto_row["Outline"]
+        cytoplasm_path = mplPath.Path(outline)
+        # check if any of the nuclei are within cytoplasm path
+        for nuc_index, nuc_row in nuc_locations.iterrows():
```
Comment (gwaybio) [inline 1.segment-data/cell-health-segmentation.py:144]: this looks like a time sink to me!

It's probably premature to optimize now (a general rule of thumb is to leave optimization for later, after a baseline implementation).

In the future, if we do need to optimize, this is a good candidate place to start.

File: 1.segment-data/test.ipynb [hunk 17]
```diff
@@ -0,0 +1,126 @@
+{
```
Comment (gwaybio) [inline 1.segment-data/test.ipynb]: maybe this file added by mistake?

**Review by gwaybio — APPROVED** [review]
Minor comments, LGTM!

File: 1.segment-data/cell-health-segmentation.py (hunk 16 above)
Comment (roshankern) [reply 1.segment-data/cell-health-segmentation.py:144]: Slightly optimized with code below. More changes to do in the future.

```
# drop nucleus so it doesn't have to be rechecked
nuc_locations.drop(nuc_index, inplace=True)
```
